archunit/src/main/java/com/tngtech/archunit/library/GeneralCodingRules.java

```diff
@@ -478,6 +483,14 @@ public void check(JavaClass implementationClass, ConditionEvents events) {
                     });
                 }
             }
+
+            private boolean allPossibleTestClassesHaveImplementationInRightPackage(List<JavaClass> possibleTestClasses) {
+                return possibleTestClasses.stream()
+                        .allMatch(testClazz -> classesByPackageName.getOrDefault(testClazz.getPackageName(), emptyList())
+                                .stream()
+                                .filter(clazz -> !testClazz.getSimpleName().equals(clazz.getSimpleName()))
+                                .anyMatch(clazz -> testClazz.getSimpleName().startsWith(clazz.getSimpleName())));
```

Comment:
We could check the exact expected test name (to avoid false negatives when the test name coincidentally starts with the name of another unrelated class): 

```suggestion
                                 .anyMatch(clazz -> testClazz.getSimpleName().equals(clazz.getSimpleName() + testClassSuffix)));
```

(which requires passing `testClassSuffix` as another method parameter)